### PR TITLE
NodeInfo response should use a collection rather than fields

### PR DESCRIPTION
**Pull request by williamrandolph**
When the nodes info endpoint is pluggable, the `NodeInfo` objects will need to be able to hold more than just a known group of heterogeneous objects representing different kinds of information. To this end, this PR introduces a `ReportingService.Info` interface, which each of the component info objects now inherits. The NodeInfo class can then replace its individual fields with a collection of `ReportingService.Info` objects. For this collection, I've chosen a map keyed on object type, which gives us the ability to handle casting dynamically in the `getInfo()` method. I'm open to alternate approaches.

I'm leaving the serialization of the `NodeInfo` object alone until I'm sure about how it should look in the long term. We don't want to get stuck supporting an intermediate serialization in future releases. (See #54305 for context.)

Relates #52975

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java [hunk 1]
```diff
@@ -147,60 +129,22 @@ public Settings getSettings() {
...
+    public <T extends ReportingService.Info> T getInfo(Class<T> clazz) {
+        return clazz.cast(infoMap.get(clazz));
     }
 
     @Nullable
     public ByteSizeValue getTotalIndexingBuffer() {
         return totalIndexingBuffer;
     }
 
+    private void addInfoIfNonNull(Class<? extends ReportingService.Info> clazz, ReportingService.Info info) {
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java]: Maybename it `maybeAddInfo(...)`? I'm unsure if that's an improvement.

**Review by pugnascotia — APPROVED** [review]
I'm not mad keen on using classes as a keys, but I'm struggling to think of an alternative, so LGTM.

**Comment by williamrandolph** [comment]
The alternative I can think of is just returning a `ReportingService.Info` object and letting the client code handle the casting if it needs to.

The strategy for heterogeneous containers that I've chosen here is discussed in _Effective Java, 2nd Edition_, item 29: "Consider typesafe heterogeneous containers".

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java [hunk 2]
```diff
@@ -147,60 +129,22 @@ public Settings getSettings() {
...
-
-    @Nullable
-    public PluginsAndModules getPlugins() {
-        return this.plugins;
-    }
-
-    @Nullable
-    public IngestInfo getIngest() {
-        return ingest;
+    // TODO: javadoc on the utility of a method that hides casting
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java:132]: Can this java doc be added before committing?

**Review by rjernst — APPROVED** [review]
This is a clever approach. I think it will make using the new response a lot easier extracting a typed info. The one caveat is it means we need to ensure each info type class is unique when registering services with infos, but that is for later. LGTM!

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java (hunk 2 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodeInfo.java]: Javadoc added!

**Comment by williamrandolph** [comment]
@elasticmachine please run elasticsearch-ci/2

**Pull request merged**